larsim/PhotonPropagation/PhotonVisibilityService.cc

```diff
@@ -306,7 +313,85 @@ namespace phot {
+        for (geo::TPCGeo const& TPC : cryo.IterateTPCs()) {
+          tpcVolume.ExtendToInclude(TPC.ActiveBoundingBox());
+        }
+      }
+
+      // Try to make ~10cm voxels, fit TPC exactly, then fit cryostat approximately
+      if (fUseAutomaticVoxels) {
+
+        std::string logString = "Automatic voxelisation x-dimension\n";
+        findVoxelSuggestion(logString,
+                            tpcVolume.MinX(),
+                            tpcVolume.MaxX(),
+                            cryoVolume.MinX(),
+                            cryoVolume.MaxX(),
+                            fNx,
+                            fXmin,
+                            fXmax);
+        logString += "Automatic voxelisation y-dimension\n";
+        findVoxelSuggestion(logString,
+                            tpcVolume.MinY(),
+                            tpcVolume.MaxY(),
+                            cryoVolume.MinY(),
+                            cryoVolume.MaxY(),
+                            fNy,
+                            fYmin,
+                            fYmax);
+        logString += "Automatic voxelisation z-dimension\n";
+        findVoxelSuggestion(logString,
+                            tpcVolume.MinZ(),
+                            tpcVolume.MaxZ(),
+                            cryoVolume.MinZ(),
+                            cryoVolume.MaxZ(),
+                            fNz,
+                            fZmin,
+                            fZmax);
+        mf::LogInfo("PhotonVisibilityService") << logString;
+      }
+
       fVoxelDef = sim::PhotonVoxelDef(fXmin, fXmax, fNx, fYmin, fYmax, fNy, fZmin, fZmax, fNz);
+
+      // Output lists of voxels within the TPC and outside it
+      if (fSaveTPCVoxels != "" || fSaveOtherVoxels != "") {
+
+        unsigned int const totalVoxels = fVoxelDef.GetNVoxels();
+        std::vector<unsigned int> tpcVoxels, otherVoxels;
+        tpcVoxels.reserve(totalVoxels);
+        otherVoxels.reserve(totalVoxels);
+        for (unsigned int voxelID = 0; voxelID < totalVoxels; ++voxelID) {
+
+          auto const voxelCenter = fVoxelDef.GetPhotonVoxel(voxelID).GetCenter();
+          if (tpcVolume.ContainsPosition(voxelCenter)) { tpcVoxels.push_back(voxelID); }
+          else {
+            otherVoxels.push_back(voxelID);
+          }
+        }
+        std::string logString = "Voxels within TPC: " + std::to_string(tpcVoxels.size());
+        if (fSaveTPCVoxels != "") logString += " saved to file\n\t" + fSaveTPCVoxels;
+        logString += "\nVoxels outside TPC: " + std::to_string(otherVoxels.size());
+        if (fSaveOtherVoxels != "") logString += " saved to file\n\t" + fSaveOtherVoxels;
+        mf::LogInfo("PhotonVisibilityService") << logString << std::endl;
+
+        if (fSaveTPCVoxels != "") {
+          std::ofstream outputFile = std::ofstream(fSaveTPCVoxels);
+          for (auto voxel : tpcVoxels)
+            outputFile << std::to_string(voxel) + "\n";
```

Comment:
No need to convert voxel to string before streaming to the output file.

```suggestion
            outputFile << voxel << '\n';
```

Reply:
Oops, been spending too much time with Python. Fixed.